src/IRMatch.h

```diff
@@ -552,9 +552,27 @@ Const pattern_arg(int64_t x) {
 }
 HALIDE_ALWAYS_INLINE
 const SpecificExpr pattern_arg(const Expr &e) {
-    return {*e.get()};
+    return {e};
+}
+
+// Helpers to deref SpecificExprs to const BaseExprNode & rather than
+// passing them by value anywhere (incurring lots of refcounting)
+template<typename T,
+         // T must be a pattern node
+         typename = typename std::decay<T>::type::pattern_tag,
+         // But T may not be SpecificExpr
+         typename = typename std::enable_if<!std::is_same<typename std::decay<T>::type, SpecificExpr>::value>::type>
```

Comment:
(Unrelated: On the one hand, I'm glad that C++11, etc lets us do this sort of thing, but geez is the syntax painful.)